### ISSUE TITLE
Add IME support for each platform

### ISSUE THREAD

**Issue by daipom** [glfw/glfw#2097]
Hi, we are currently developing functions about IME support.

* https://github.com/clear-code/glfw/tree/3.4-2022-03-31+im-support

The main goal is to enable comfortable Japanese input on each platform by adding APIs to manage IME such as setting the candidate window position.
If possible, we would like to develop a form that can be merged into the upstream GLFW.

We are still in the early stage of implementation and will continue.
We would like to hear your view on the direction from this stage.

The functions we are developing are based on the PR: https://github.com/glfw/glfw/pull/658
The goal of this PR is the same as ours.

The APIs are currently the same as this PR and they are as follows.

* About preedit
    * `glfwSetPreeditCallback`: callback to notify the preedit text
    * `glfwSetPreeditCursorPos`: set the candidate window position
    * `glfwGetPreeditCursorPos`: get the candidate window position
    * `glfwResetPreeditText`: reset the preedit text
* About IME status
    * `glfwSetIMEStatusCallback`: callback to notify the change of ON/OFF of the IME
    * `glfwSetInputMode`: switch ON/OFF of the IME by using new mode `GLFW_IME`
    * `glfwGetInputMode`: get ON/OFF of the IME by using new mode `GLFW_IME`

We added the following fixes to the PR.

* Apply to the latest GLFW
* Fix the code styles and add some refactoring
* Add some improvements
    * Win32: use dynamic load for Imm32
    * Win32: enable to commit the results to the input
    * X11: use over-the-spot by default because on-the-spot of XIM doesn't work well
    * macOS: enable to manage the candidate window position
    * macOS: enable to clear preedit after the commit

# Details for each platform

You can confirm the behavior by `tests/ime`.
Right click to set that position to the position of the candidate window.

We are also developing a sample application using [raylib](https://github.com/clear-code/raylib/tree/better-ime-support), a videogames programing library using GLFW.

* https://github.com/clear-code/RaylibIMEInputSampleApp

## Win32

* Use Imm32.
* All APIs work fine on Windows10.

`tests/ime`

![glfw_test_ime_win](https://user-images.githubusercontent.com/12229857/166201487-8b2209f9-b7ba-4e96-9620-38e710ed5dfa.png)

Sample application

![sample_app_win](https://user-images.githubusercontent.com/12229857/166201505-52e0bf44-4dd2-4c5e-9ea4-8df701b6c145.png)

## X11

* Environment: Ubuntu 20.04 LTS, X11, ibus
* Adopted over-the-spot style by default because on-the-spot style of XIM doesn't work well.
    * You can change the style to on-the-spot by setting the `IM_STYLE` environmental variable `on-the-spot`.
* Over-the-spot style
    * Only `glfwSetPreeditCursorPos` and `glfwGetPreeditCursorPos` work and other APIs don't work.
    * This is OK for this style because the input method side manages the preedit, so the application side doesn't need manage it except for the candidate window position.
* On-the-spot style
    * `glfwSetPreeditCursorPos` and `glfwGetPreeditCursorPos` don't work, so we can't manage the candidate window position.
    * All APIs of IME status don't work too.

`tests/ime`

![glfw_test_ime_x11_overthespot](https://user-images.githubusercontent.com/12229857/166201551-b5d3c067-8fd3-411b-817e-7f59f9a7e617.png)

Sample application

![sample_app_x11_overthespot](https://user-images.githubusercontent.com/12229857/166201566-13adee6f-4e40-4223-8923-0a7a56458172.png)

## macOS

* All APIs work fine on M1 BigSur 11.3.
* There is a problem with the slow display of the candidate window.

`tests/ime`

<img width="604" alt="glfw_test_ime_mac" src="https://user-images.githubusercontent.com/12229857/166201624-fdfff570-b040-45a7-a469-4bdc0d94efb8.png">

Sample application

<img width="804" alt="sample_app_mac" src="https://user-images.githubusercontent.com/12229857/166201631-910e7b65-6164-4a3b-8e42-fd9944380d89.png">

# Future considerations

We will continue to improve and organize detailed implementation.

Apart from that, we would like to discuss the entire future direction for merging.

## Windows: TSF support

In addition to Imm32, we need to consider supporting TSF (Text Service Framework).

[A comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-992132924) pointed out that the TSF needs to be supported instead of Imm32.

However, Imm32 seems to work well for now, so it may be enough to support only Imm32 at first?

Can we think of a direction to have this merged into upstream without TSF support, and then consider TSF later if necessary?

## GNU/Linux: Wayland support

In addition to XIM, we need to consider supporting Wayland.

There is an idea supporting Fcitx directly in [a comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-165163366).
However, we are not considering this direction because over-the-spot style seems to work well on XIM and we can support Wayland if needed.

What we would like to discuss is whether there is any possibility to go in the direction of only supporting XIM at first?

## Fullscreen mode

[A comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-992132924) also pointed to the following:

> there should be a way to get candidate list, as input method can't show its candidate list window when the game is in fullscreen mode.

There is indeed a problem with the candidate window not displaying properly in fullscreen mode.

However, it seems to be impossible to get the candidate list on some platforms such as macOS, so we think it is difficult to go in the direction of adding a API to get the candidate list and having the application side draw it

Instead of that method, we are considering somehow making it possible to display the candidate window in full screen mode as well, but nothing has been figured out yet. What do you think about this direction?

## Detailed IME status APIs

[A comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-992132924) also pointed to the following:

> there should be callbacks to tell what input method is, and which mode(Native/AlphaNumeric) the input method currently in.

We think that we don't need to consider this kind of API first, since these APIs are not mandatory for the application side and are not easy to implement well on each platform and in each language.

We think the current API for ON/OFF of the IME should be sufficient at first. What do you think about this?

**Comment by daipom** [comment]
We want to develop a form that can be merged into the upstream GLFW.

So we would be happy to discuss the above `Future considerations` and other requirements.

**Comment by ashie** [comment]
> ## Windows: TSF support
> 
> In addition to Imm32, we need to consider supporting TSF (Text Service Framework).
> 
> [A comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-992132924) pointed out that the TSF needs to be supported instead of Imm32.
> 
> However, Imm32 seems to work well for now, so it may be enough to support only Imm32 at first?
> 
> Can we think of a direction to have this merged into upstream without TSF support, and then consider TSF later if necessary?

I'm working on this task with @daipom

TSF is nice to have, but we decided that we won't include it in our first pull request.
Here is the reason:

* Recent Windows have a compatible layer "CUAS" (Cicero Unaware Application Support: TSF is formerly known as Cinero) that converts TSF <-> Imm32, so that we can use TSF input methods from Imm32 applications.
  * CUAS is enabled by default on Windows Vista or later.
  * Although old CUAS was buggy (at least on Windows XP), recent version of CUAS seems stable enough.
* Although Windows 8's Metro style didn't allow Imm32, Windows 10's tablet mode seems supporting it. We confirmed the actual behavior.
* Although UWP applications can't use Imm32, It's not concerned with GLFW becuase GLFW is a library which depends on Win32.
* We confirmed actual behavior with various input methods of Windows 10, and most of them works fine.
  * Microsoft IME (Japanese)
  * ATOK Passport (Japanese)
  * Google Japanese Input
  * Microsoft Pinyin (Simplified Chinese)
  * Microsoft Bopomofo (Traditional Chinese)
  * Chinese Traditional Array (Traditional Chinese)
* We don't use TSF specific features yet.
* Since TSF requires COM, we can't build it by MinGW.
  * Even when we use Visual Studio, we have to greatly change the project constitution.

**Comment by ashie** [comment]
> ## GNU/Linux: Wayland support
> 
> In addition to XIM, we need to consider supporting Wayland.
> 
> There is an idea supporting Fcitx directly in [a comment on the original PR](https://github.com/glfw/glfw/pull/658#issuecomment-165163366). However, we are not considering this direction because over-the-spot style seems to work well on XIM and we can support Wayland if needed.
> 
> What we would like to discuss is whether there is any possibility to go in the direction of only supporting XIM at first?

Wayland support is a bit complicated but we think we need to support it. Because recent major distributions are migrating from X11 to Wayland, and Wayland doesn't have global positioning protocol or API, so that it's always needed to support Wayland's text input protocol to show candidate window in correct position. Supporting Fcitx or IBus directly doesn't resolve this issue.

An complicated issue of Wayland is that there are some protocols and each of them are incompatible:

* [text-input-v1](https://cgit.freedesktop.org/wayland/wayland-protocols/tree/unstable/text-input/text-input-unstable-v1.xml)
* [text-input-v3](https://cgit.freedesktop.org/wayland/wayland-protocols/tree/unstable/text-input/text-input-unstable-v3.xml)
* [gtk-text-input](https://github.com/GNOME/gtk/blob/eb4d736a4dea14528d51475d97bbf013fb81571a/modules/input/gtk-text-input.xml)

I'm not sure but it seems that gtk-text-input doesn't seem used in recent days (It's not included in GTK4, probably dropped).
To support 2 major desktop environment "GNOME" and "KDE" on GNU/Linux, probably supporting text-input-v3 is enough.
But another major window system "Weston" is also used widely on embedded systems and it supports only text-input-v1.
So we may need to support both text-input-v1 and text-input-v3.

We have experience to implement them (e.g. https://github.com/ashie/gtk-wayland-textinputv1) so that we can implement it in GLFW as well, but not yet started to work on it.

**Comment by daipom** [comment]
We created PR #2117, so now we close this issue.
